lib/dice/src/cert.rs

```diff
@@ -0,0 +1,346 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{
+    alias_cert_tmpl, deviceid_cert_tmpl, SerialNumber, NOTBEFORE_LENGTH,
+};
+use hubpack::SerializedSize;
+use salty::constants::{
+    PUBLICKEY_SERIALIZED_LENGTH, SIGNATURE_SERIALIZED_LENGTH,
+};
+use salty::signature::{Keypair, PublicKey};
+use serde::{Deserialize, Serialize};
+use serde_big_array::BigArray;
+use unwrap_lite::UnwrapLite;
+
+#[derive(Debug)]
+pub enum CertError {
+    BadSig,
+    NoPubKey,
+    NoSig,
+    NoSignData,
+    TooSmall,
+    NotFound,
+    NoCn,
+}
+
+pub trait Cert {
+    fn as_bytes(&self) -> &[u8];
+    fn as_mut_bytes(&mut self) -> &mut [u8];
+
+    const SERIAL_NUMBER_START: usize;
+    const SERIAL_NUMBER_END: usize;
+
+    fn get_serial_number(&self) -> u8 {
+        u8::from_be_bytes(
+            self.as_bytes()[Self::SERIAL_NUMBER_START..Self::SERIAL_NUMBER_END]
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    fn set_serial_number(mut self, sn: u8) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::SERIAL_NUMBER_START..Self::SERIAL_NUMBER_END]
+            .copy_from_slice(&sn.to_be_bytes());
+
+        self
+    }
+
+    const ISSUER_SN_START: usize;
+    const ISSUER_SN_END: usize;
+
+    fn get_issuer_sn(&self) -> SerialNumber {
+        SerialNumber::from_bytes(
+            &self.as_bytes()[Self::ISSUER_SN_START..Self::ISSUER_SN_END]
+                .try_into()
+                .unwrap_lite(),
+        )
+    }
+
+    fn set_issuer_sn(mut self, sn: &SerialNumber) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::ISSUER_SN_START..Self::ISSUER_SN_END]
+            .copy_from_slice(sn.as_bytes());
+
+        self
+    }
+
+    const NOTBEFORE_START: usize;
+    const NOTBEFORE_END: usize;
+
+    fn set_notbefore(mut self, utctime: &[u8; NOTBEFORE_LENGTH]) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::NOTBEFORE_START..Self::NOTBEFORE_END]
+            .copy_from_slice(utctime);
+
+        self
+    }
+
+    const SUBJECT_SN_START: usize;
+    const SUBJECT_SN_END: usize;
+
+    fn get_subject_sn(&self) -> SerialNumber {
+        SerialNumber::from_bytes(
+            &self.as_bytes()[Self::SUBJECT_SN_START..Self::SUBJECT_SN_END]
+                .try_into()
+                .unwrap_lite(),
+        )
+    }
+
+    fn set_subject_sn(mut self, sn: &SerialNumber) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::SUBJECT_SN_START..Self::SUBJECT_SN_END]
+            .copy_from_slice(sn.as_bytes());
+
+        self
+    }
+
+    const PUB_START: usize;
+    const PUB_END: usize;
+
+    fn get_pub(&self) -> &[u8] {
+        &self.as_bytes()[Self::PUB_START..Self::PUB_END]
+    }
+
+    fn set_pub(mut self, pubkey: &[u8; PUBLICKEY_SERIALIZED_LENGTH]) -> Self
```

Comment:
We could. This case is a bit weird since this `_LENGTH` is coming from a different library. But either way `_START`, `_END`, `_LENGTH` stuff from the `*_tmpl.rs` is mostly generated at this point so the `_LENGTH` values in these files are just calculated values from `_START` and `_END` behind the scenes.